crates/re_types_builder/src/codegen/cpp/mod.rs

```diff
@@ -1426,11 +1426,73 @@ fn quote_fill_arrow_array_builder(
                     let arrow_builder_type = arrow_array_builder_type(&variant.typ, objects);
                     let variant_name = format_ident!("{}", variant.name);
 
-                    let variant_append = if variant.typ.is_plural() {
-                        let error = format!("Failed to serialize {}::{}: list types in unions not yet implemented", obj.name, variant.name); // TODO(#2919)
-                        quote! {
-                            (void)#variant_builder;
-                            return Error(ErrorCode::NotImplemented, #error);
+                    let variant_append = if let Some(element_type) = variant.typ.plural_inner() {
+                        if variant.is_nullable {
+                            let error = format!("Failed to serialize {}::{}: nullable list types in unions not yet implemented", obj.name, variant.name);
+                            quote! {
+                                (void)#variant_builder;
+                                return Error(ErrorCode::NotImplemented, #error);
+                            }
+                        } else if arrow_builder_type == "ListBuilder" {
+                            let field_name = format_ident!("{}", variant.snake_case_name());
+
+                            if *element_type == ElementType::Float16 {
+                                // We need an extra cast for float16:
+                                quote! {
+                                    ARROW_RETURN_NOT_OK(variant_builder->Append());
+                                    auto value_builder =
+                                        static_cast<arrow::HalfFloatBuilder *>(variant_builder->value_builder());
+                                    const rerun::half* halfs = union_instance._data.#field_name.data();
+                                    ARROW_RETURN_NOT_OK(value_builder->AppendValues(
+                                        reinterpret_cast<const uint16_t*>(halfs),
+                                        static_cast<int64_t>(union_instance._data.#field_name.size())
+                                    ));
+                                }
+                            } else {
+                                let type_builder_name = match element_type {
```

Comment:
any reason you can't re-use `arrow_array_builder_type`?

Reply:
yes - we have an `element_type`, and `arrow_array_builder_type` takes a `Type`